### ISSUE TITLE
RangeError: Maximum call stack size exceeded

### ISSUE THREAD

**Issue by avi-dwave** [eslint/eslint#13867]
**Tell us about your environment**

* **Node version:** v10.23.0
* **npm version:** v6.14.8
* **Local ESLint version:** v7.13.0
* **Global ESLint version:** v7.13.0

**What parser (default, `@babel/eslint-parser`, `@typescript-eslint/parser`, etc.) are you using?**
@babel/eslint-parser

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```
parser: "@babel/eslint-parser"

extends:
  - airbnb
  - plugin:import/errors
  - plugin:import/warnings
  - prettier
  - prettier/react

plugins:
  - eslint-plugin-no-inline-styles

settings:
  import/resolver:
    babel-module: {}

env:
  browser: true
  jest: true

parserOptions:
  ecmaVersion: 2018
  sourceType: module

rules:
  quotes: [error, double]
  max-classes-per-file: error
  import/no-absolute-path: 0
  import/no-dynamic-require: 0
  global-require: 0
  react/jsx-props-no-spreading: 0
  react/forbid-prop-types: [2, { forbid: [any, array] }]
  no-restricted-syntax: ["error", "WithStatement"]
  no-inline-styles/no-inline-styles: 2

```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

import Flyout from "/component/base/Flyout/Flyout";
import Meter from "/component/base/Meter";
import SpacedContent from "/component/base/SpacedContent";
import Grid from "/component/base/Grid";
import Text from "/component/base/Text";
import Link from "/component/base/Link";
import Separator from "/component/base/Separator";
import Icon from "/component/base/Icon";

import FetchContext from "/component/fetch/FetchContext";

/* Styled Components */
const Row = styled.div`
  padding: 0 1rem;
`;

const ProjectNameRow = styled.div`
  margin-left: ${(props) => (24 * props.depth).toString(10)}px;
  color: var(--midnight);
  font-weight: 600;
  margin-bottom: 16px;
  cursor: default;
  word-wrap: anywhere;
`;

const ProjectLinkRow = styled(ProjectNameRow)`
  position: relative;
  display: flex;
  align-items: center;
  color: ${(props) => (props.isActiveProject ? "blue" : "var(--grey)")};
  font-weight: normal;
  .check-icon {
    margin-left: 4px;
    font-size: 18px;
  }
`;
/* /Styled Components */

const ProjectNode = ({
  project,
  nodes,
  depth,
  setActiveProject,
  activeProject,
}) => {
  const isActiveProject = project.name === activeProject.name;
  if (nodes && nodes !== undefined && nodes.length > 0) {
    return (
      <>
        <ProjectNameRow depth={depth}>
          <div className="flex flex-between items-center">
            <div>{project.name}</div>
          </div>
        </ProjectNameRow>
        {nodes.map((node) => (
          <ProjectNode
            key={node.project.name}
            project={node.project}
            nodes={node.children}
            depth={depth + 1}
            setActiveProject={setActiveProject}
            activeProject={activeProject}
          />
        ))}
      </>
    );
  }
  return (
    <ProjectLinkRow depth={depth} isActiveProject={isActiveProject}>
      <Link
        to="/"
        target="_self"
        hoverColor="blue"
        onClick={setActiveProject.bind(this, project.id)}
        className="flex flex-between items-center"
      >
        <div>{project.name}</div>
      </Link>
      {isActiveProject && <Icon name="check" className="check-icon" />}
    </ProjectLinkRow>
  );
};

const projectType = PropTypes.shape({
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
});

const nodesType = PropTypes.arrayOf(
  PropTypes.shape({ project: projectType, children: nodesType }),
);

ProjectNode.propTypes = {
  project: projectType.isRequired,
  nodes: nodesType,
  depth: PropTypes.number.isRequired,
  setActiveProject: PropTypes.func.isRequired,
  activeProject: projectType,
};

class ProfileFlyoutV2 extends React.PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      timeRemaining: 0,
      timeUsed: 0,
      totalTime: 0,
      planName: "",
      github: {
        username: "",
        repoURL: "",
      },
      billingMyAccountUrl: "",
      fetchingInProgress: false,
      paymentPlatformLogOutUrl: window.csAppProperties.payment.logoutURL,
      paymentAvailability: window.csAppProperties.payment.availability,
      enablePaymentPlatformLogout: false, // Enable/Disable logout of payment platform when user logging out of Leap
      activeProject: {
        name: "",
      },
      projectView: false,
      projects: [],
    };
    this.getBillingMyAccountUrl = this.getBillingMyAccountUrl.bind(this);
    this.goToBillingMyAccount = this.goToBillingMyAccount.bind(this);
    this.logOut = this.logOut.bind(this);
  }

  getGithub() {
    this.props
      .fetch("/leap/api/account/github/", {
        credentials: "include",
        method: "GET",
      })
      .then((result) => {
        this.setState({
          github: {
            username: result.data.username,
            repoURL: result.data.repoURL,
          },
        });
      });
  }

  getActiveProject() {
    this.props
      .fetch("/leap/api/account/active_project/", {
        credentials: "include",
        method: "GET",
      })
      .then((result) => {
        this.setState({
          activeProject: {
            name: result.data.project.name,
          },
        });
      });
  }

  getProjects() {
    this.props
      .fetch("/leap/api/account/projects", {
        credentials: "include",
        method: "GET",
      })
      .then((result) => {
        this.setState({
          projects: result.data.projects,
        });
      });
  }

  getBillingMyAccountUrl() {
    var data = {};
    var component = this;
    fetch("/payment/sso/billing/", {
      credentials: "include",
      method: "POST",
      body: JSON.stringify(data),
      headers: {
        Accept: "application/json",
        "X-CSRFToken":
          window && window.csAppProperties && window.csAppProperties.CSRF_TOKEN,
      },
    })
      .then(
        function (response, self) {
          return response
            .json()
            .catch(() => Promise.resolve({ redirect_url: "" }));
        },
        function (error) {},
      )
      .then(function (data, self) {
        component.setState({ billingMyAccountUrl: data.redirect_url });
      });
  }

  componentDidMount() {
    this.setState({
      fetchingInProgress: true,
    });
    this.props
      .fetch("/leap/api/account/qpu")
      .then(({ data }) => data)
      .then((data) =>
        this.setState({
          timeRemaining: data.timeRemaining,
          timeUsed: data.totalTime - data.timeRemaining,
          totalTime: data.totalTime,
          planName: data.planName,
          fetchingInProgress: false,
        }),
      );
    this.getGithub();
    this.getBillingMyAccountUrl();
    this.getActiveProject();
    this.getProjects();
  }

  getPlanName() {
    if (this.state.planName === "Developer Plan") {
      if (!this.state.github.username) {
        return "Trial Plan";
      }
    }

    return this.state.planName;
  }

  goToBillingMyAccount(e) {
    e.preventDefault();
    window.location.href = this.state.billingMyAccountUrl;
  }

  logOut(e) {
    e.preventDefault();
    window.csAppProperties.loogutInProgress = true;
    fetch("/leap/logout/", {
      method: "GET",
      headers: {
        "X-Requested-With": "XMLHttpRequest",
      },
    }).then((result) => {
      window.location.href = this.state.paymentPlatformLogOutUrl;
    });
  }

  setActiveProject(project_id, e) {
    e.preventDefault();
    fetch("/leap/api/account/active_project/", {
      credentials: "include",
      method: "PUT",
      body: JSON.stringify({ project_id: project_id }),
      headers: {
        "Content-Type": "application/json",
        "X-CSRFToken":
          window && window.csAppProperties && window.csAppProperties.CSRF_TOKEN,
      },
    }).then((result) => {
      window.location.reload();
    });
  }

  render() {
    const {
      billingMyAccountUrl,
      paymentAvailability,
      enablePaymentPlatformLogout,
      planName,
      timeRemaining,
      totalTime,
      timeUsed,
      fetchingInProgress,
      projectView,
      projects,
      activeProject,
    } = this.state;
    let timeRemainingText;
    let meterColor = "smokyBlue";
    const percentage =
      totalTime > 0 ? ((timeUsed / totalTime) * 100).toFixed(2) : 0;
    if (!planName || this.getPlanName() === "Trial Plan") {
      timeRemainingText = `Trial Plan Used ${percentage}%`;
    } else {
      timeRemainingText = `Subscription Used ${percentage}%`;
    }

    let StandardView = (
      <SpacedContent size="small" header trailer>
        <Row>
          <Grid>
            <Grid.Col span={6}>
              <Text color="midnight">{this.getPlanName()}</Text>
            </Grid.Col>
            <Grid.Col span={6}>
              <Text align="right" color="blue">
                <a href="/leap/plans" className="link">
                  Get More Time
                </a>
              </Text>
            </Grid.Col>
          </Grid>
        </Row>
        <Row>
          <Meter value={timeUsed / totalTime} color={meterColor} />
        </Row>
        <Row>
          {!fetchingInProgress ? (
            timeRemaining === null ? (
              "Unlimited time remaining this month"
            ) : totalTime ? (
              timeRemainingText
            ) : totalTime === 0 && !fetchingInProgress ? (
              "No time remaining this month"
            ) : (
              ""
            )
          ) : (
            <div className="loading-spinner-container">
              <div className="lds-ellipsis">
                <div></div>
                <div></div>
                <div></div>
                <div></div>
              </div>
              Loading Data...
            </div>
          )}
        </Row>
        <Separator />
        <Row>
          <Link to="/settings" hoverColor="blue" onClick={this.props.close}>
            <Icon name="gear" label color="cobalt" /> Settings
          </Link>
        </Row>
        {planName == "Commercial Plan" && billingMyAccountUrl ? (
          <Row>
            <Link
              to="#"
              hoverColor="blue"
              target="_self"
              onClick={this.goToBillingMyAccount.bind(this)}
            >
              <Icon name="billing" label color="cobalt" /> Orders and Billing
            </Link>
          </Row>
        ) : (
          <Row>
            <Link
              to="/"
              className="disabled-cursor disabled-link"
              onClick={(e) => e.preventDefault()}
            >
              <Icon name="billing-disabled" label color="cobalt" /> Orders and
              Billing
            </Link>
          </Row>
        )}
        <Row>
          <Link
            to={`${window.csAppProperties.zendeskUrl}/hc/en-us`}
            hoverColor="blue"
          >
            <Icon name="help" label color="cobalt" /> Support
          </Link>
        </Row>
        {this.state.projects.length > 1 && (
          <Row>
            <Link
              to="/"
              target="_self"
              hoverColor="blue"
              onClick={(e) => {
                e.preventDefault();
                this.setState({ projectView: true });
              }}
              className="flex flex-between items-center"
            >
              <div>
                <Icon name="tag" label color="cobalt" /> Projects
              </div>
              <Icon name="chevron-right" label />
            </Link>
          </Row>
        )}
        <Separator />
        {!paymentAvailability ? (
          <Row>
            <Link to="/logout/" hoverColor="blue" target="_self">
              <Icon name="logout" label /> Log Out
            </Link>
          </Row>
        ) : enablePaymentPlatformLogout ? (
          <Row>
            <Link
              to="/"
              hoverColor="blue"
              target="_self"
              onClick={this.logOut.bind(this)}
            >
              <Icon name="logout" label /> Log Out
            </Link>
          </Row>
        ) : (
          <Row>
            <Link to="/logout/" hoverColor="blue" target="_self">
              <Icon name="logout" label /> Log Out
            </Link>
          </Row>
        )}
      </SpacedContent>
    );

    let ProjectView = (
      <SpacedContent size="small" header trailer>
        <Row>
          <Grid>
            <Grid.Col span={6}>
              <Link
                to="/"
                target="_self"
                hoverColor="blue"
                onClick={(e) => {
                  e.preventDefault();
                  this.setState({ projectView: false });
                }}
                className="flex items-center"
              >
                <Icon name="chevron-left" label />
                <Text className="ml1" color="midnight">
                  Back
                </Text>
              </Link>
            </Grid.Col>
          </Grid>
        </Row>
        <Separator />
        {projects.map((project, index) => {
          return (
            <ProjectNode
              key={index}
              project={project.project}
              nodes={project.children || []}
              depth={1}
              setActiveProject={this.setActiveProject}
              activeProject={activeProject}
            />
          );
        })}
      </SpacedContent>
    );

    return (
      <Flyout {...this.props}>
        <Text size="small" color="cobalt">
          {projectView ? ProjectView : StandardView}
        </Text>
      </Flyout>
    );
  }
}

export default (props) => (
  <FetchContext.Consumer>
    {(fetch) => <ProfileFlyoutV2 fetch={fetch} {...props} />}
  </FetchContext.Consumer>
);

```

```bash
eslint ./assets/js/shared/component/partial/ProfileFlyoutV2/ProfileFlyoutV2.jsx
```

**What did you expect to happen?**
eslint would show me any linting issues with my code

**What actually happened? Please include the actual, raw output from ESLint.**
```bash
Oops! Something went wrong! :(

ESLint: 7.13.0

RangeError: Maximum call stack size exceeded
Occurred while linting /home/developer/work/sapi/qubist/metalab-cloud-service/assets/js/shared/component/partial/ProfileFlyoutV2/ProfileFlyoutV2.jsx:97
    at Array.join (native)
    at iterateProperties (/home/developer/work/sapi/qubist/metalab-cloud-service/node_modules/eslint-plugin-react/lib/util/propTypes.js:445:55)
    at iterateProperties (/home/developer/work/sapi/qubist/metalab-cloud-service/node_modules/eslint-plugin-react/lib/util/propTypes.js:65:7)
    at buildReactDeclarationTypes (/home/developer/work/sapi/qubist/metalab-cloud-service/node_modules/eslint-plugin-react/lib/util/propTypes.js:443:11)
    at buildReactDeclarationTypes (/home/developer/work/sapi/qubist/metalab-cloud-service/node_modules/eslint-plugin-react/lib/util/propTypes.js:458:25)
    at iterateProperties (/home/developer/work/sapi/qubist/metalab-cloud-service/node_modules/eslint-plugin-react/lib/util/propTypes.js:446:29)
    at iterateProperties (/home/developer/work/sapi/qubist/metalab-cloud-service/node_modules/eslint-plugin-react/lib/util/propTypes.js:65:7)
    at buildReactDeclarationTypes (/home/developer/work/sapi/qubist/metalab-cloud-service/node_modules/eslint-plugin-react/lib/util/propTypes.js:443:11)
    at buildReactDeclarationTypes (/home/developer/work/sapi/qubist/metalab-cloud-service/node_modules/eslint-plugin-react/lib/util/propTypes.js:458:25)
    at iterateProperties (/home/developer/work/sapi/qubist/metalab-cloud-service/node_modules/eslint-plugin-react/lib/util/propTypes.js:446:29)
```

**Are you willing to submit a pull request to fix this bug?**
maybe

**Comment by g-plane** [comment]
Hi, according to your information, it seems that this is an issue from "eslint-plugin-react", not ESLint itself. You can open issue at its repository.
